laurent22/joplin
# Chore: Refactor warning banner out of `ScreenHeader` into a separate file

# Summary

This pull request moves the warning logic out of `ScreenHeader.tsx` and into a new `WarningBanner.tsx` component.

This pull request is related to https://github.com/laurent22/joplin/issues/9658 and should make adding an "accept share" warning box (like the one on desktop) easier.

# Notes

At present, `components/ScreenHeader.tsx` is moved to `components/ScreenHeader/index.tsx`. The new `WarningBanner` component, being a sub-component of the screen header, is created in `components/ScreenHeader/`.

# Testing

This pull request has an automated test, but it also needs to be tested manually to do that:
1. Sync with Joplin Server
2. Set the maximum item size to a small number, add a note larger than that, and re-sync.
3. Verify that the "some items cannot be synchronised" banner is visible.
4. Verify that clicking on it navigates to the sync status screen.
5. Apply the following diff:
	```diff
	diff --git a/packages/lib/Synchronizer.ts b/packages/lib/Synchronizer.ts
	index 3ba2ea086..e7d05e5d8 100644
	--- a/packages/lib/Synchronizer.ts
	+++ b/packages/lib/Synchronizer.ts
	@@ -508,6 +508,7 @@ export default class Synchronizer {
	 					// Note: that's probably not needed anymore?
	 					// await uploadSyncInfo(this.api(), remoteInfo);
	 				}
	+				throw { code: 'outdatedSyncTarget' };
	 			} catch (error) {
	 				if (error.code === 'outdatedSyncTarget') {
	 					Setting.setValue('sync.upgradeState', Setting.SYNC_UPGRADE_STATE_SHOULD_DO);
	```
6. Sync
7. Verify that a "the sync target needs to be upgraded" button is visible: ![3eef244186f408ea36c34accc007cc93.png](:/deda04d7e39a45caa3477352bca9e7bb)
8. Click on the upgrade button
9. Verify that the upgrade button isn't visible while upgrading the sync target.
10. Un-apply the patch.
11. (At around this point, I changed the sync target)
12. Enable encryption & sync
13. Apply the following diff (to trigger a decryption error):
	```diff
	diff --git a/packages/lib/services/e2ee/EncryptionService.ts b/packages/lib/services/e2ee/EncryptionService.ts
	index 109e765d4..a55358866 100644
	--- a/packages/lib/services/e2ee/EncryptionService.ts
	+++ b/packages/lib/services/e2ee/EncryptionService.ts
	@@ -655,7 +655,7 @@ export default class EncryptionService {
	 	}
	 
	 	public isValidEncryptionMethod(method: EncryptionMethod) {
	-		return [EncryptionMethod.SJCL, EncryptionMethod.SJCL1a, EncryptionMethod.SJCL1b, EncryptionMethod.SJCL2, EncryptionMethod.SJCL3, EncryptionMethod.SJCL4].indexOf(method) >= 0;
	+		return [EncryptionMethod.SJCL, EncryptionMethod.SJCL1b, EncryptionMethod.SJCL2, EncryptionMethod.SJCL3, EncryptionMethod.SJCL4].indexOf(method) >= 0;
	 	}
	 
	 	public async itemIsEncrypted(item: any) {
	```
14. Add a new note on another Joplin client and sync.
15. From the patched version of Joplin, sync 2-3 times.
16. Verify that the "some items cannot be decrypted" warning is visible and links to the sync status screen.

This has been tested successfully on an Android 12 emulator.
